sdk/search/Azure.Search.Documents/src/Indexes/FieldBuilder.cs

```diff
@@ -0,0 +1,456 @@
+                    [typeof(double)] = SearchFieldDataType.Double,
+                    [typeof(bool)] = SearchFieldDataType.Boolean,
+                    [typeof(DateTime)] = SearchFieldDataType.DateTimeOffset,
+                    [typeof(DateTimeOffset)] = SearchFieldDataType.DateTimeOffset,
+#if EXPERIMENTAL_SPATIAL
+                    [typeof(PointGeometry)] = SearchFieldDataType.GeographyPoint,
+#else
+                    [typeof(GeographyPoint)] = SearchFieldDataType.GeographyPoint,
+#endif
+                });
+
+        private static readonly ISet<Type> s_unsupportedTypes =
+            new HashSet<Type>
+            {
+                typeof(decimal),
+            };
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="FieldBuilder"/> class.
+        /// </summary>
+        public FieldBuilder()
+        {
+        }
+
+        /// <summary>
+        /// Gets or sets the <see cref="ObjectSerializer"/> to use to generate field names that match JSON property names.
+        /// You should use hte same value as <see cref="SearchClientOptions.Serializer"/>.
+        /// <see cref="JsonObjectSerializer"/> will be used if no value is provided.
+        /// </summary>
+        public ObjectSerializer Serializer { get; set; }
+
+        /// <summary>
+        /// Creates a list of <see cref="SearchField"/> objects corresponding to
+        /// the properties of the type supplied.
+        /// </summary>
+        /// <param name="modelType">
+        /// The type for which fields will be created, based on its properties.
+        /// </param>
+        /// <returns>A collection of fields.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="modelType"/>.</exception>
+        public IList<SearchField> Build(Type modelType)
+        {
+            Argument.AssertNotNull(modelType, nameof(modelType));
+
+            ArgumentException FailOnNonObjectDataType()
+            {
+                string errorMessage =
+                    $"Type '{modelType}' does not have properties which map to fields of an Azure Search index. Please use a " +
+                    "class or struct with public properties.";
+
+                throw new ArgumentException(errorMessage, nameof(modelType));
+            }
+
+            Serializer ??= new JsonObjectSerializer();
+            IMemberNameConverter nameProvider = Serializer as IMemberNameConverter ?? DefaultSerializedNameProvider.Shared;
+
+            if (ObjectInfo.TryGet(modelType, nameProvider, out ObjectInfo info))
+            {
+                if (info.Properties.Length == 0)
+                {
+                    throw FailOnNonObjectDataType();
+                }
+
+                // Use Stack to avoid a dependency on ImmutableStack for now.
+                return Build(modelType, info, nameProvider, new Stack<Type>(new[] { modelType }));
+            }
+
+            throw FailOnNonObjectDataType();
```

Comment:
Would it simplify things to remove the extra block scoped method like?

```C#
            if (ObjectInfo.TryGet(modelType, nameProvider, out ObjectInfo info) &&
                info.Properties.Length > 0)
            {
                // Use Stack to avoid a dependency on ImmutableStack for now.
                return Build(modelType, info, nameProvider, new Stack<Type>(new[] { modelType }));
            }

             string errorMessage =
                    $"Type '{modelType}' does not have properties which map to fields of an Azure Search index. Please use a " 
                    "class or struct with public properties.";
                throw new ArgumentException(errorMessage, nameof(modelType));
```

Reply:
This was ported track 1 code with minimal changes desired.